src/main/java/com/limechain/trie/structure/nibble/Nibbles.java

```diff
@@ -92,191 +84,148 @@ public boolean startsWith(Nibbles prefix) {
-    }
-
-    @Override
-    public boolean remove(Object o) {
-        return nibbles.remove(o);
-    }
-
-    @Override
-    public boolean containsAll(@NotNull Collection<?> c) {
-        return new HashSet<>(nibbles).containsAll(c);
-    }
-
-    @Override
-    public boolean addAll(@NotNull Collection<? extends Nibble> c) {
-        return nibbles.addAll(c);
-    }
-
-    @Override
-    public boolean addAll(int index, @NotNull Collection<? extends Nibble> c) {
-        return nibbles.addAll(c);
-    }
-
-    @Override
-    public boolean removeAll(@NotNull Collection<?> c) {
-        return nibbles.removeAll(c);
+    private static String toLowerHexString(Iterable<Nibble> nibbles) {
+        return StreamSupport.stream(nibbles.spliterator(), false)
+            .map(Nibble::asLowerHexDigit)
+            .collect(
+                Collector.of(
+                    StringBuilder::new,
+                    StringBuilder::append,
+                    StringBuilder::append,
+                    StringBuilder::toString
+                )
+            );
     }
 
-    @Override
-    public boolean retainAll(@NotNull Collection<?> c) {
-        return nibbles.retainAll(c);
+    /**
+     * Adds a new nibble to the nibbles.
+     * @param nibble the new nibble to add
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(nibble);
+        return new Nibbles(newNibbles);
     }
 
-    @Override
-    public void clear() {
-        nibbles.clear();
+    /**
+     * Adds a new nibble to the nibbles at the given position.
+     * @param nibble the new nibble to add
+     * @return the new Nibbles after insertion
+     * @implNote will create a new modified copy of the nibbles, does not mutate the instance invoked on
+     */
+    public Nibbles add(int index, Nibble nibble) {
+        List<Nibble> newNibbles = new ArrayList<>(this.nibbles);
+        newNibbles.add(index, nibble);
+        return new Nibbles(newNibbles);
     }
 
-    @Override
+    /**
+     * Returns the element at the specified position in this list.
+     *
+     * @param index index of the element to return
+     * @return the element at the specified position in this list
+     * @throws IndexOutOfBoundsException if the index is out of range
+     *         ({@code index < 0 || index >= size()})
+     */
     public Nibble get(int index) {
         return nibbles.get(index);
     }
 
-    @Override
-    public Nibble set(int index, Nibble element) {
-        return nibbles.get(index);
+    // NOTE:
+    //  Since Nibbles is immutable, we don't actually need to explicitly copy
+    //  This method could go away, passing by reference could suffice, although going back to mutability
+    //  if later decided would become rather difficult.
```

Comment:
That's an interesting conundrum, take a look at the note.

Reply:
Why would mutability become difficult to switch to?

Reply:
Because someone would have to track the ownership anew, which is currently given by `copy`. Everywhere we now copy, the new owner would want to own the data and probably mutate it at some point. These mutations generate new instances now, which is the easy part to refactor (when switching to mutability of Nibbles) - make them mutate the old instance, but that would cause changes in previous owners too, since by removing `copy` we will have passed by reference only.

Reply:
We decided we'll leave `copy` as a shallow implementation returning `this` only for explicitness and traceability of ownership's sake.